File: src/0173-Binary-Search-Tree-Iterator/0173.cpp
#include <stack>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();

class BSTIterator 
{
public:
    BSTIterator(TreeNode* root) 
    {
        pushLeft(root);
    }
    
    /** @return the next smallest number */
    int next() 
    {
        auto node = small.top(); small.pop();
        pushLeft(node->right);
        return node->val;
    }
    
    /** @return whether we have a next smallest number */
    bool hasNext() 
    {
        return small.empty() ? false : true;
    }
    
private:
    stack<TreeNode *> small;
    
    void pushLeft(TreeNode* node)
    {
        while (node)
        {
            small.push(node);
            node = node->left;
        }
    }
};

File: src/0173-Binary-Search-Tree-Iterator/0173.py
class BSTIterator:
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = list()
        self.pushLeft(root)
        
    def next(self):
        """
        @return the next smallest number
        :rtype: int
        """
        node = self.stack.pop()
        self.pushLeft(node.right)
        return node.val
        
    def pushLeft(self, node):
        while node:
            self.stack.append(node)
            node = node.left
        
    def hasNext(self):
        """
        @return whether we have a next smallest number
        :rtype: bool
        """
        return True if self.stack else False